core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,263 @@
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toLong(), unit)
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(longProgression.step(safeMultiply(value, unit.days.toLong())))
+
+/**
+ * Creates a [LocalDateProgression] from `this` down to [that], inclusive.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.simpleRangeCreation
+ */
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression].
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @throws IllegalArgumentException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.random(random: Random = Random) : LocalDate = longProgression.random(random).let(LocalDate.Companion::fromEpochDays)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression] or null if the progression is empty.
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.randomOrNull(random: Random = Random) : LocalDate? = longProgression.randomOrNull(random)
+    ?.let(LocalDate.Companion::fromEpochDays)
+
+internal fun LongProgression.random(random: Random = Random) : Long = random.nextLong(0L..(last - first) / step) * step + first
+
+internal fun LongProgression.randomOrNull(random: Random = Random) : Long? = if (isEmpty()) null else random(random)
+
+internal fun LongProgression.contains(element: Long) : Boolean = element in (if(step > 0) first..last else last..first) && (element - first) % step == 0L
+
+internal val LongProgression.size: Int
```

Comment:
This looks to be an issue because the negative of `Long.MIN_VALUE` wraps back around to itself due to an overflow error. Do we want to solve this by implementing a `safeSubtract` method instead of using `safeAdd` on the negative of the subtrahend, or is it not worth the trouble, and we can get by with this explanatory comment?